crates/mun_compiler/src/diagnostics.rs

```diff
@@ -3,111 +3,510 @@ use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
 use mun_syntax::{ast, AstNode, SyntaxKind};
 use std::cell::RefCell;
 
-mod emit;
+use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
 
-pub use emit::Emit;
-use mun_errors::{Diagnostic, Level};
+use unicode_segmentation::UnicodeSegmentation;
+
+/// Replace all tabs('\t') in string to spaces(' ')
+fn replace_tabs(text: &mut String) {
+    let mut tab_index = text.find('\t');
+    while let Some(index) = tab_index {
+        #[allow(clippy::range_plus_one)]
+        text.replace_range(index..index + 1, " ");
+        tab_index = text.find('\t');
+    }
+}
+
+/// Returns amount of graphemes in text range
+fn graphemes_amount_in_range(range: (usize, usize), text: &str) -> usize {
+    let (_, text) = text.split_at(range.0);
+    let (text, _) = text.split_at(range.1 - range.0);
+    UnicodeSegmentation::graphemes(text, true).count()
+}
+
+/// Constructs proper range for `Snippet` from `annotate-snippets` crate with taking
+/// folded text part in account, by subtraction first line offset from output range
+fn construct_range(
+    highliht_range: (usize, usize),
```

Comment:
Maybe use `Range<usize>` instead?

Reply:
> Maybe use `Range<usize>` instead?

`(usize, usize)` is what `annotate-snippets` api want for constructing `Snippet`.

Reply:
Strange, but OK :)